### PR TITLE
Remove FSDP wrapping from sub-models.

### LINKED ISSUE #34113: fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP is not working with the Trainer
### System Info

image
acc_cfg.yml:

compute_environment: LOCAL_MACHINE
debug: false
distributed_type: FSDP
downcast_bf16: 'no'
enable_cpu_affinity: true
fsdp_config:
  fsdp_activation_checkpointing: true
  fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
  fsdp_backward_prefetch: NO_PREFETCH
  fsdp_cpu_ram_efficient_loading: true
  fsdp_forward_prefetch: true
  fsdp_offload_params: true
  fsdp_sharding_strategy: FULL_SHARD
  fsdp_state_dict_type: SHARDED_STATE_DICT
  fsdp_sync_module_states: true
  fsdp_use_orig_params: true
machine_rank: 0
main_process_ip: 0.0.0.0
main_process_port: 0
main_training_function: main
mixed_precision: bf16
num_machines: 3
num_processes: 24
rdzv_backend: etcd-v2
same_network: false
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

accelerate launch --config_file acc_cfg.yml train.py $TRAINING_ARGS
the train.py is any training script that train using transformers.Trainer
$TRAINING_ARGS are the TrainingArguments plus some path to data
fdsp_trans

### Expected behavior

Train Paligemma model with FSDP and have PaliGemmaMultiModalProjector wrapped.

### PR DESCRIPTION
**Pull request by eljandoubi**
# What does this PR do?

Fixes #34113 

## Who can review?

Library:

- trainer: @muellerzr and @SunMarc

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -2258,6 +2259,10 @@ def _inner_training_loop(
         # FSDP-XLA, SageMaker MP/DP, DataParallel, IPEX
         use_accelerator_prepare = True if model is self.model else False
 
+        if use_accelerator_prepare and self.is_fsdp_enabled:
+            # Remove FSDP wrapping from sub-models.
+            self.model = extract_model_from_parallel(self.model, recursive=True)
+
```
Comment (SunMarc) [inline src/transformers/trainer.py:2265]: You can use `unwarp_model` function in transformers instead. Also, why do we need to set `recursive` to `True` ? Also, please leave a comment above as this specific path is only to make it functional with `auto_batch_size`.

**Review by SunMarc — COMMENTED** [review]
Thnaks for fixing the issue @eljandoubi ! Do you think there is a simpler way to handle this edge case @muellerzr ?

File: src/transformers/trainer.py (hunk 1 above)
Comment (eljandoubi) [reply src/transformers/trainer.py:2265]: Unwrap_model do not give  access to recursive argument. The auto wrap policies wrap sub modules with fsdp and unwrap_model is unable to remove it. Y can try to apply unwrap_model on the toy example from https://pytorch.org/tutorials/intermediate/FSDP_tutorial.html for rank=0 and world_size = 1. Then experiment with the line I have written.

**Comment by eljandoubi** [comment]
@SunMarc @muellerzr Did you get a different result than I did?

**Review by muellerzr — APPROVED** [review]
Thanks for the fix, can you add a test in `tests/test_trainer.py` for this?

File: src/transformers/trainer.py (hunk 1 above)
Comment (SunMarc) [reply src/transformers/trainer.py]: I'm talking about this [function](https://github.com/huggingface/transformers/blob/ef976a7e181b78abf2f1ba7ea02e506ea1cb111e/src/transformers/modeling_utils.py#L5464) in transformers. It uses extract_model_from_parallel under the hood so it should be comparable.

**Review by SunMarc — APPROVED** [review]
Thanks ! Left an suggestion for `unwrap_model`

File: src/transformers/trainer.py (hunk 1 above)
Comment (eljandoubi) [reply src/transformers/trainer.py]: Euh I see.

**Comment by eljandoubi** [comment]
@SunMarc I migrated to `unwrap_model`.

**Review by LysandreJik — APPROVED** [review]
Let's merge it if you're both ok with it @SunMarc @muellerzr

**Comment by SunMarc** [comment]
Please rebase this PR on main in order to pass the CI @eljandoubi !

**Comment by eljandoubi** [comment]
@SunMarc @LysandreJik @muellerzr Is there any update on the pull request?

**Comment by ArthurZucker** [comment]
We were on a company wide offsite! merging as they all approved 🤗

**Pull request merged**